service/src/main/java/org/cbioportal/service/util/ProfiledCasesCounter.java

```diff
@@ -58,32 +59,42 @@ public void calculate(List<String> molecularProfileIds, List<String> sampleIds,
                 .map(x -> x.getStudyId() + (countByPatients ? x.getPatientId() : x.getSampleId()))
                 .collect(Collectors.toSet());
 
+        
+        Map<String, Set<String>> genePanelToCases = new HashMap<String, Set<String>>();
+        
         for (AlterationCountByGene alterationCountByGene : alterationCounts) {
             final Set<String> profiledCasesForGene = new HashSet<String>();
             Integer entrezGeneId = alterationCountByGene.getEntrezGeneId();
             List<GenePanel> allPanels = new ArrayList<>();
             
+            Set<String> totalProfiledCases = new HashSet<String>();
+            Set<String> allMatchingGenePanelIds = new HashSet<String>();
+            // different calculations depending on if gene is linked to gene panels 
             if (geneGenePanelMap.containsKey(entrezGeneId)) {
-                geneGenePanelMap.get(entrezGeneId).forEach(genePanel -> {
-                    Set<String> casesWithPanelData = genePanelDataMap
+                // calculate and store cases for each gene panel
+                for (GenePanel genePanel : geneGenePanelMap.get(entrezGeneId)) {
```

Comment:
Updated `genePanelToCases` can be completely outside of block iterating `alterationCounts` (line 63).

may be
```
        Map<String, Set<String>> genePanelToCases = genePanelDataMap
                .entrySet().stream()
                .collect(Collectors
                        .toMap(
                                e -> e.getKey(),
                                e -> {
                                    return e.getValue()
                                            .stream()
                                            .map(x -> x.getStudyId() + (countByPatients ? x.getPatientId() : x.getSampleId()))
                                            .collect(Collectors.toSet());
                                    }
                                ));
```

Reply:
I'm still trying to intuitively understand these constructs.
My interpretation of your suggested code:
GenePanelData records are linkages between a sample/patient/profile/gene_panel/study and a boolean flag for whether or not the sample was "profiled". I don't fully understand how a sample would be part of a genetic profile and yet not be profiled ... although I guess I know that there are some samples which are part of a mutation profile and which were profiled but for which no mutation events were found. I still think there is some constraint in this collection of objects which I do not understand.
Anyway, what we fetch via fetchGenePanelDataInMultipleMolecularProfiles is a big aggregate of all of these membership records, and then we group these into buckets based on the gene panel id. So the map genePanelDataMap has one bucket of sample memberships per panel. These can span across profiles and studies.
So the suggested code will loop through every entry in genePanelDataMap (each panel) and insert a mapping from the panel_id to a set of "sample/study" identifier strings, which will be constructed by looping through each of the membership records which were in the bucket for that gene panel, and aggregating / uniqueifying the cases.
Once this is done, we would not need to check within the loop which examines all genes whether or not the genePanelToCases entry for this gene panel needs to be computed or not --- because we have precomputed all of them.

Ok, I think I understand.